wsnet/listen.go

```diff
@@ -316,3 +336,58 @@ func (l *listener) Close() error {
 func (l *listener) Addr() net.Addr {
 	return nil
 }
+
+// normalizeHost converts all representations of "localhost" to "localhost".
+func normalizeHost(addr string) string {
+	ip := net.ParseIP(addr)
+	if ip == nil {
+		return addr
+	}
+
+	if localNet.Contains(ip) {
+		return "localhost"
+	}
+	return addr
+}
+
+// getAddress parses the data channel's protocol into an address suitable for
+// net.Dial. It also verifies that the ProtoMessage permits connecting to said
+// address.
+func getAddress(msg ProtoMessage, protocol string) (netwk, addr string, err error) {
+	parts := strings.SplitN(protocol, ":", 3)
+	if len(parts) != 3 {
+		return "", "", fmt.Errorf("invalid dial address: %v", protocol)
+	}
+
+	var (
+		network  = parts[0]
+		host     = normalizeHost(parts[1])
+		port     = parts[2]
+		fullAddr = net.JoinHostPort(host, port)
+	)
+	if len(msg.Policies) == 0 {
+		return network, fullAddr, nil
+	}
+
+	portParsed, err := strconv.Atoi(port)
+	if err != nil || portParsed < 0 || bits.Len(uint(portParsed)) > 16 {
+		return "", "", fmt.Errorf("invalid dial address %q port: %v", protocol, port)
+	}
+	portParsedU16 := uint16(portParsed)
+
+	for _, p := range msg.Policies {
+		if p.Network != "" && p.Network != network {
+			continue
+		}
+		if p.Host != "" && normalizeHost(p.Host) != host {
+			continue
+		}
+		if p.Port != 0 && p.Port != portParsedU16 {
+			continue
+		}
+
+		return network, fullAddr, nil
+	}
+
+	return "", "", fmt.Errorf("connections are not permitted to %q", err)
```

Comment:
It'd be nice if we had this as an exposed Err... exactly like we do for the `Op` err type.

Something like `ErrNoPermission`

Reply:
added a new type `notPermittedByPolicyErr` and a op `OpNotPermittedByPolicy = "port_not_permitted"`